pkg/agent/protocol/protocol.go

```diff
@@ -100,17 +94,36 @@ func (d *disruptor) Apply(ctx context.Context, duration time.Duration) error {
 		wc <- d.proxy.Start()
 	}()
 
-	if err := d.redirector.Start(); err != nil {
-		return fmt.Errorf(" failed traffic redirection: %w", err)
-	}
-
 	// On termination, restore traffic and stop proxy
 	defer func() {
-		// ignore errors when stopping. Nothing to do
-		_ = d.redirector.Stop()
 		_ = d.proxy.Stop()
 	}()
 
+	if d.config.Transparent {
+		trCfg := iptables.TrafficRedirectorConfig{
+			Executor: d.executor,
+		}
+		// Redirect traffic to the proxy
+		tr := &iptables.TrafficRedirectionSpec{
+			Iface:           d.config.Iface,
+			DestinationPort: d.config.TargetPort,
+			RedirectPort:    d.config.RedirectPort,
+		}
+
+		redirector, err := iptables.NewTrafficRedirectorWithConfig(tr, trCfg)
+		if err != nil {
+			return err
+		}
+
+		if err := redirector.Start(); err != nil {
+			return fmt.Errorf(" failed traffic redirection: %w", err)
+		}
+
+		defer func() {
+			_ = redirector.Stop()
+		}()
+	}
+
```

Comment:
Another alternative to this, would be to make the redirector an interface, and have two implementations: `iptables.Redirector` and a `NoopRedirector` that does nothing. Then, instead of the big `if` block here, we could pass the `disruptor` a `NoopRedirector` if we're just proxying and an `iptables.Redirector` if we're running in transparent mode.

Dropping this here as an alternative approach in case you want to consider it. I don't have strong concerns against the simple `if`.

Reply:
This makes sense. I will give it a try.

Reply:
On second thought I will make it in a follow-up PR so as not to mix concerns here.